File: Capítulo_13/Algoritmos de busca/Busca por interpolação/interpolationSearch.js
import {
    defaultCompare,
    Compare,
    defaultEquals,
    defaultDiff,
    biggerOrEquals,
    lesserOrEquals,
} from "../../Models/util.js";
const DOES_NOT_EXIST = -1;

// Algoritmo interpolation search
function interpolationSearch(
    array,
    value,
    compareFn = defaultCompare,
    equalsFn = defaultEquals,
    diffFn = defaultDiff
) {
    const { length } = array;
    let low = 0;
    let high = length - 1;
    let position = -1;
    let delta = -1;
    while (
        low <= high &&
        biggerOrEquals(value, array[low], compareFn) &&
        lesserOrEquals(value, array[high], compareFn)
    ) {
        delta = diffFn(value, array[low]) / diffFn(array[high], array[low]);
        position = low + Math.floor((high - low) * delta);
        if (equalsFn(array[position], value)) {
            return position;
        }
        if (compareFn(array[position], value) === Compare.LESS_THAN) {
            low = position + 1;
        } else {
            high = position - 1;
        }
    }
    return DOES_NOT_EXIST;
}

// Testando a função interpolationSearch
let array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
let value = 4;
console.log(interpolationSearch(array, value)); // 3

File: Capítulo_13/Algoritmos de busca/Busca por interpolação/Código em Python/interpolationSearch.py
# Compare elements
Compare = {
    "LESS_THAN": -1,
    "BIGGER_THAN": 1,
    "EQUALS": 0
}

# defaultCompare
def default_compare(a, b):
    if a == b:
        return Compare["EQUALS"]
    return Compare["LESS_THAN"] if a < b else Compare["BIGGER_THAN"]

# Função defaultEquals
def default_equals(a, b):
    return a == b

# Função default_diff
def default_diff(a,  b):
    return float(a) - float(b)

# Função create_non sorted array: cria um array de valores não ordenados
def create_non_sorted_array(size):
    array = []
    for i in range(size, 0, -1):
        array.append(i)

    return array

# Variável DOES_NOT_EXIST
DOES_NOT_EXIST = -1

# Função lesser_or_equals
def lesser_or_equals(a, b, compare_fn):
    comp = compare_fn(a, b)
    return comp == Compare["LESS_THAN"] or comp == Compare["EQUALS"]

# Função bigger_or_equals
def bigger_or_equals(a, b, compare_fn):
    comp = compare_fn(a, b)
    return comp == Compare["BIGGER_THAN"] or comp == Compare["EQUALS"]